src/c-writer.cc

```diff
@@ -0,0 +1,2437 @@
+        } else {
+          Write(", ");
+          if ((count % 8) == 0)
+            Write(Newline());
+        }
+
+        Write(DefineLocalScopeName(index_to_name[local_index]), " = 0");
+        ++count;
+      }
+      ++local_index;
+    }
+    if (count != 0) {
+      Dedent(4);
+      Write(";", Newline());
+    }
+  }
+}
+
+void CWriter::WriteStackVarDeclarations() {
+  for (Type type : {Type::I32, Type::I64, Type::F32, Type::F64}) {
+    size_t count = 0;
+    for (const auto& pair: stack_var_sym_map_) {
+      Type stp_type = pair.first.second;
+      const std::string& name = pair.second;
+
+      if (stp_type == type) {
+        if (count == 0) {
+          Write(type, " ");
+          Indent(4);
+        } else {
+          Write(", ");
+          if ((count % 8) == 0)
+            Write(Newline());
+        }
+
+        Write(name);
+        ++count;
+      }
+    }
+    if (count != 0) {
+      Dedent(4);
+      Write(";", Newline());
+    }
+  }
+}
+
+void CWriter::Write(const ExprList& exprs) {
+  for (const Expr& expr: exprs) {
+    switch (expr.type()) {
+      case ExprType::Binary:
+        Write(*cast<BinaryExpr>(&expr));
+        break;
+
+      case ExprType::Block: {
+        const Block& block = cast<BlockExpr>(&expr)->block;
+        std::string label = DefineLocalScopeName(block.label);
+        size_t mark = MarkTypeStack();
+        PushLabel(LabelType::Block, block.label, block.sig);
+        Write(block.exprs, LabelDecl(label));
+        ResetTypeStack(mark);
+        PopLabel();
+        PushTypes(block.sig);
+        break;
+      }
+
+      case ExprType::Br:
+        Write(GotoLabel(cast<BrExpr>(&expr)->var), Newline());
+        // Stop processing this ExprList, since the following are unreachable.
+        return;
+
+      case ExprType::BrIf:
+        Write("if (", StackVar(0), ") {");
+        DropTypes(1);
+        Write(GotoLabel(cast<BrIfExpr>(&expr)->var), "}", Newline());
+        break;
+
+      case ExprType::BrTable: {
+        const auto* bt_expr = cast<BrTableExpr>(&expr);
+        Write("switch (", StackVar(0), ") ", OpenBrace());
+        DropTypes(1);
+        Index i = 0;
+        for (const Var& var : bt_expr->targets)
+          Write("case ", i++, ": ", GotoLabel(var), Newline());
+        Write("default: ");
+        Write(GotoLabel(bt_expr->default_target), Newline(), CloseBrace(),
+              Newline());
+        // Stop processing this ExprList, since the following are unreachable.
+        return;
+      }
+
+      case ExprType::Call: {
+        const Var& var = cast<CallExpr>(&expr)->var;
+        const Func& func = *module_->GetFunc(var);
+        Index num_params = func.GetNumParams();
+        Index num_results = func.GetNumResults();
+        assert(type_stack_.size() >= num_params);
+        if (num_results > 0) {
+          assert(num_results == 1);
+          Write(StackVar(num_params - 1, func.GetResultType(0)), " = ");
+        }
+
+        Write(GlobalVar(var), "(");
```

Comment:
it appears that there is a problem with this line, because the following code

```WebAssembly
(module
  (func $fac (param f64) (result f64)
    get_local 0
    f64.const 1
    f64.lt
    if (result f64)
      f64.const 1
    else
      get_local 0
      get_local 0
      f64.const 1
      f64.sub
      call $fac
      f64.mul
    end)
  (export "fac" (func $fac)))
```

is compiled as

```c
static f64 fac(f64 _) {
  FUNC_PROLOGUE;
  u32 i0;
  f64 d0, d1, d2;
  d0 = _;
  d1 = 1;
  i0 = d0 < d1;
  if (i0) {
    d0 = 1;
  } else {
    d0 = _;
    d1 = _;
    d2 = 1;
    d1 -= d2;
    d1 = (d1);
    d0 *= d1;
  }
  FUNC_EPILOGUE;
  return d0;
}
```